### PR TITLE
Refactor salts into separate file. Better docstrings.

### PR DESCRIPTION
**Pull request by lucasvo**

### CODE REVIEW

File: proofs/salts.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+		} else if saltsType.Kind() == reflect.Slice {
+			err = handleFillSaltsSlice(saltsType, saltsField, valueField)
+		} else if saltsType.Kind() == reflect.Struct {
+			err = handleFillSaltsStruct(saltsField, saltsType, valueField)
+		} else {
+			return fmt.Errorf("Invalid type (%s) for field", reflect.TypeOf(saltsField.Interface()).String())
+		}
+
+		if err != nil {
+			return
```
Comment (vimukthi-git) [inline proofs/salts.go]: Why are we not returning the error here?

File: proofs/salts.go [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+			checkType = reflect.TypeOf(saltsFieldItem.Interface())
+		}
+
+		if checkType.Kind() == reflect.Struct {
+			err = handleFillSaltsStruct(saltsFieldItem, checkType, dataFieldItem)
+		} else {
+			err = handleFillSaltsValue(reflect.Indirect(saltsFieldItem))
+		}
+
+		if err != nil {
```
Comment (vimukthi-git) [inline proofs/salts.go]: same here?

File: proofs/salts_test.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
+package proofs
+
+import (
+	"github.com/centrifuge/precise-proofs/examples/documents"
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+func TestFillSalts(t *testing.T) {
```
Comment (vimukthi-git) [inline proofs/salts_test.go]: I think it would be nicer to separate out the cases in this test into multiple tests

File: proofs/tree.go [hunk 4]
```diff
@@ -52,16 +52,17 @@ import (
...
 
-/*
-Allows to pass certain options to the tree generation logic. Currently support:
-	1. EnableHashSorting: Implement a merkle tree with sorted hashes
-	2. SaltsLengthSuffix: As precise proofs support repeated fields, when generating the merkle tree we need to add a leaf that represents the length of the slice.
-	The default suffix is `Length`, although it is customizable so it does not collide with potential field names of your own proto structs.
-*/
+// TreeOptions allows customizing the generation of the tree
 type TreeOptions struct {
+	//	EnableHashSorting: Implement a merkle tree with sorted hashes
```
Comment (vimukthi-git) [inline proofs/tree.go]: 👍

File: proofs/salts.go (hunk 1 above)
Comment (lucasvo) [reply proofs/salts.go]: @mikiquantum do you know why? I just moved this code from A to B.

Comment (mikiquantum) [reply proofs/salts.go]: Error is implicitly returned by named returns.

File: proofs/salts.go (hunk 2 above)
Comment (mikiquantum) [reply proofs/salts.go]: Same reason as above.

**Pull request merged**